keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+
+UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
+
+### Stop using component config internal types
+
+As kubeadm will no longer default, validate or migrate component configs inside its code base, the need to use component config internal types is no longer founded. In that case the usage of internal types can be removed. This would also allow us to reduce the number of imported packages and to drop one of the last dependencies on the kubernetes/kubernetes code base.  
+The switch is going to be performed by replacing each internal type with a corresponding versioned type. This, by itself, is not going to have any impact on users.
+
+### “kubeadm upgrade plan” changes
+As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, an ahead of time notice about what actions are required is a nice thing to have.  
+Hence, the “kubeadm upgrade plan” command is going to provide information on the component configs. In particular, if they are managed by kubeadm or are user supplied, and if they need manual upgrade or not.
+
+### Risks and Mitigations
+
+#### Users will have to manually migrate more Kubernetes core component configs than before
+
+In general, most of the users are expected to change only a few fields in the core Kubernetes component configs.
+Thus, such configs are better managed by maintaining a Kustomize patch set, rather than a full blown component config.  
+With a proper documentation in place, users will be encouraged to opt in for more kubeadm generated configs and Kustomize patch sets, when necessary.
+
+#### Users will have to manually migrate more AddOn component configs than before
+
+For user supplied AddOn configs, that have changed config versions, this might be somewhat true.  
+In the long run, addon operators are expected to manage component configs.
+
+#### Users will loose track of what needs manual migration and what not
+
+The proposed changes to the "kubeadm upgrade plan" command will mitigate this risk, as kubeadm users are required to run that command prior to performing an actual upgrade.
+
+## Design Details
+
+### Test Plan
+
+In addition to unit tests, extensive E2E tests should be created to verify various aspects if the implementation.  
+Most notably E2E tests should be written to verify different aspects of:
+- AddOn and Kubernetes core component configs
+- Configs supplied by users or generated by kubeadm
+- behavior during kubeadm init and upgrade
```

Comment:
just noting that if the components require manual conversion of component configs we won't be able to have tests for that, until the components support `--convert`.

i see the potential for a test similar to https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-kustomize-master that exercises the patching aspect of component config.